### ISSUE TITLE
Connect - Discover tab not discovering local Jolokia JVM agents

### ISSUE THREAD

**Issue by tadayosi** [hawtio/hawtio#3320]
Discover tab has been working fine, but now I noticed that it no longer works and cannot detect any locally running Camel apps.

![image](https://github.com/hawtio/hawtio/assets/156692/305bd86b-188f-4217-9e5d-58c23d63877c)

I initially thought it might be related to jolokia/jolokia#667 but it doesn't seem so, as Hawtio cannot detect it whether it's Jolokia 1.7.x or 2.0.x that's attached to the Camel app. Probably [hawtio-local-jvm-mbean](https://github.com/hawtio/hawtio/tree/4.x/plugins/hawtio-local-jvm-mbean) is broken now.

**Comment by tadayosi** [comment]
It appears that the issue is at Camel side (remote Jolokia agent). What I tested:

- Hawtio 4.0-SNAPSHOT -> Camel 4.0.4 (Jolokia 1.7.2 agent): OK
- Hawtio 4.0-SNAPSHOT -> Camel 4.5.0-SNAPSHOT (Jolokia 2.0.2 agent): FAILED

When it fails, I see the following messages in the Hawtio JBang log:
```
[pool-3-thread-2] INFO org.eclipse.jetty.server.handler.ContextHandler.hawtio - jolokia-agent: /127.0.0.1--> Sending
[pool-3-thread-1] INFO org.eclipse.jetty.server.handler.ContextHandler.hawtio - jolokia-agent: /192.168.86.28--> Sending
[pool-3-thread-1] INFO org.eclipse.jetty.server.handler.ContextHandler.hawtio - jolokia-agent: /192.168.86.28--> Timeout
[qtp1224347463-28] INFO org.eclipse.jetty.server.handler.ContextHandler.hawtio - jolokia-agent: --> Couldnt send discovery message from /127.0.0.1: java.net.BindException: Can't assign requested address
[qtp1224347463-28] INFO org.eclipse.jetty.server.handler.ContextHandler.hawtio - jolokia-agent: --> Exception during lookup: java.util.concurrent.ExecutionException: org.jolokia.service.discovery.MulticastUtil$CouldntSendDiscoveryPacketException: Can't send discovery UDP packet from /127.0.0.1: Can't assign requested address
```

@grgrzybek Can you see why it fails to connect to a remote Jolokia 2.0.2 agent?

**Comment by tadayosi** [comment]
> - Hawtio 4.0-SNAPSHOT -> Camel 4.0.4 (Jolokia 1.7.2 agent): OK
> - Hawtio 4.0-SNAPSHOT -> Camel 4.5.0-SNAPSHOT (Jolokia 2.0.2 agent): FAILED

In any case, a connection from `Remote` tab with explicit connection settings works just fine. Only discovery at `Discover` tab fails.

**Comment by grgrzybek** [comment]
@tadayosi do you have some instructions about how to run sample Camel JBang application(s) to check this issue?

**Comment by tadayosi** [comment]
@grgrzybek Sure. First build the latest camel project locally. Then

```
jbang app install -Dcamel.jbang.version=4.5.0-SNAPSHOT --name camelx camel@apache/camel
jbang app install -Dcamel.jbang.version=4.0.4 --name camel40 camel@apache/camel
```

Do each in a different terminal.
```
camelx init hello1.java
camelx run hello1.java
```

```
camelx jolokia hello1
```

```
camel40 init hello2.java
camel40 run hello2.java
```

```
camel40 jolokia hello2
```

Open Hawtio with `hawtio` and go to Connect -> Discover and try to auto detect the running Camel apps.

**Comment by grgrzybek** [comment]
With the above (only for Camel 4.5.0-SNAPSHOT), I'm getting this in `camel run Hello1.java:
```
Failed to find Agent-Class manifest attribute from /home/ggrzybek/.m2/repository/org/jolokia/jolokia-agent-jvm/2.0.1/jolokia-agent-jvm-2.0.1.jar
```

**Comment by tadayosi** [comment]
@grgrzybek It's a known issue. That's why we first need to build camel. Please use `camelx run Hello1.java` instead of `camel run Hello1.java` which would make you use the released version of camel 4.4.1.

**Comment by grgrzybek** [comment]
I already did it some time ago.

First, in my working copy of Camel 4.5.0-SNAPSHOT (`main` branch), I have this (so I can debug with the source from my own repo):
```
diff --git a/dsl/camel-jbang/camel-jbang-main/dist/CamelJBang.java b/dsl/camel-jbang/camel-jbang-main/dist/CamelJBang.java
index 7d2787f085c6..e1335b270bc9 100755
--- a/dsl/camel-jbang/camel-jbang-main/dist/CamelJBang.java
+++ b/dsl/camel-jbang/camel-jbang-main/dist/CamelJBang.java
@@ -19,8 +19,8 @@
 
 //JAVA 17+
 //REPOS central=https://repo1.maven.org/maven2,apache-snapshot=https://repository.apache.org/content/groups/snapshots/
-//DEPS org.apache.camel:camel-bom:${camel.jbang.version:4.4.1}@pom
-//DEPS org.apache.camel:camel-jbang-core:${camel.jbang.version:4.4.1}
+//DEPS org.apache.camel:camel-bom:${camel.jbang.version:4.5.0-SNAPSHOT}@pom
+//DEPS org.apache.camel:camel-jbang-core:${camel.jbang.version:4.5.0-SNAPSHOT}
 //DEPS org.apache.camel.kamelets:camel-kamelets:${camel-kamelets.version:4.4.1}
 
 package main;
diff --git a/dsl/camel-jbang/camel-jbang-main/src/main/jbang/main/CamelJBang.java b/dsl/camel-jbang/camel-jbang-main/src/main/jbang/main/CamelJBang.java
index 7d2787f085c6..e1335b270bc9 100755
--- a/dsl/camel-jbang/camel-jbang-main/src/main/jbang/main/CamelJBang.java
+++ b/dsl/camel-jbang/camel-jbang-main/src/main/jbang/main/CamelJBang.java
@@ -19,8 +19,8 @@
 
 //JAVA 17+
 //REPOS central=https://repo1.maven.org/maven2,apache-snapshot=https://repository.apache.org/content/groups/snapshots/
-//DEPS org.apache.camel:camel-bom:${camel.jbang.version:4.4.1}@pom
-//DEPS org.apache.camel:camel-jbang-core:${camel.jbang.version:4.4.1}
+//DEPS org.apache.camel:camel-bom:${camel.jbang.version:4.5.0-SNAPSHOT}@pom
+//DEPS org.apache.camel:camel-jbang-core:${camel.jbang.version:4.5.0-SNAPSHOT}
 //DEPS org.apache.camel.kamelets:camel-kamelets:${camel-kamelets.version:4.4.1}
 
 package main;
```

Then I have few tweaks in `~/.jbang`:

**`~/.jbang/implicit-catalog.json`** is:
```json
{
  "catalogs": {
    "jbangdev": {
      "catalog-ref": "https://github.com/jbangdev/jbang-catalog/blob/HEAD/jbang-catalog.json"
    },
    "apache/camel": {
      "catalog-ref": "https://github.com/apache/camel/blob/HEAD/jbang-catalog.json"
    },
    "hawtio/hawtio": {
      "catalog-ref": "https://github.com/hawtio/hawtio/blob/HEAD/jbang-catalog.json"
    }
  },
  "aliases": {},
  "templates": {}
}
```

**`~/.jbang/jbang-catalog.json`** is (again - so I can use local copies of JBang `hawtio` and `camel` scripts:
```json
{
  "catalogs": {
    "camel": {
      "catalog-ref": "/data/sources/github.com/apache/camel-upstream/jbang-catalog.json"
    },
    "hawtio": {
      "catalog-ref": "/home/ggrzybek/sources/github.com/hawtio/hawtio/jbang-catalog.json"
    }
  },
  "aliases": {},
  "templates": {}
}
```

Of course `~/.jbang/bin` is on my `CLASSPATH`.

With hawtio jbang running, I see proper `JVM` discovery:
![image](https://github.com/hawtio/hawtio/assets/306565/b2c47e69-0be2-4a0e-855c-0853c223c69f)

Now checking the `Agents`...

**Comment by grgrzybek** [comment]
When you go to `http://127.0.0.1:8778/jolokia/`, you'll see:
```json
{
  "request": {
    "type": "version"
  },
  "value": {
    "agent": "2.0.2",
    "protocol": "7.3",
    "details": {
      "agent_version": "2.0.2",
      "agent_id": "49bd8238-jvm",
      "secured": false,
      "url": "http://127.0.0.1:8778/jolokia/"
    },
    "id": "49bd8238-jvm",
    "config": {
      "maxDepth": "15",
      "discoveryEnabled": "false",
      "agentId": "49bd8238-jvm",
      "maxCollectionSize": "0",
      "debug": "false",
      "agentContext": "/jolokia",
      "historyMaxEntries": "10",
      "maxObjects": "0",
      "debugMaxEntries": "100"
    },
    "info": {
      "proxy": {},
      "jmx": {}
    }
  },
  "status": 200,
  "timestamp": 1711542635
}
```

so (Multicast UDP) discovery is not enabled. I'm not sure how to enabled it with `camel jolokia`...

**Comment by grgrzybek** [comment]
Finally with:
```
diff --git a/dsl/camel-jbang/camel-jbang-core/src/main/java/org/apache/camel/dsl/jbang/core/commands/process/Jolokia.java b/dsl/camel-jbang/camel-jbang-core/src/main/java/org/apache/camel/dsl/jbang/core/commands/process/Jolokia.java
index fd9292db0ddc..147924b05e0b 100644
--- a/dsl/camel-jbang/camel-jbang-core/src/main/java/org/apache/camel/dsl/jbang/core/commands/process/Jolokia.java
+++ b/dsl/camel-jbang/camel-jbang-core/src/main/java/org/apache/camel/dsl/jbang/core/commands/process/Jolokia.java
@@ -62,7 +62,7 @@ public class Jolokia extends ProcessBaseCommand {
             } else {
                 // find a new free port to use when starting a new connection
                 long port = AvailablePortFinder.getNextAvailable(8778, 10000);
-                options = new OptionsAndArgs(null, "--port", Long.toString(port), "start", Long.toString(pid));
+                options = new OptionsAndArgs(null, "--port", Long.toString(port), "--discoveryEnabled", "true", "start", Long.toString(pid));
             }
             VirtualMachineHandlerOperations vmHandler = PlatformUtils.createVMAccess(options);
             CommandDispatcher dispatcher = new CommandDispatcher(options);
```

I can see discovery enabled and this:
![image](https://github.com/hawtio/hawtio/assets/306565/d7fa8662-b8eb-4b7a-b5e3-228bf9bad5e6)

Don't forget (if you're on Fedora with restrictive firewall settings):
```
# nft add rule inet firewalld filter_IN_FedoraServer_allow ip daddr 239.192.48.84 accept
```

**Comment by tadayosi** [comment]
Great investigation. Why was it automatically enabled with Jolokia 1.7.2 agent?  I don’t think Camel has changed so much when upgrading it to 2.0.x.

**Comment by grgrzybek** [comment]
For Camel 4.0.4 I see:
```json
{
  "request": {
    "type": "version"
  },
  "value": {
    "agent": "1.7.1",
    "protocol": "7.2",
    "config": {
      "listenForHttpService": "true",
      "maxCollectionSize": "0",
      "authIgnoreCerts": "false",
      "agentId": "192.168.0.221-68028-8ada132-jvm",
      "agentType": "jvm",
      "policyLocation": "classpath:/jolokia-access.xml",
      "agentContext": "/jolokia",
      "mimeType": "text/plain",
      "discoveryEnabled": "true",
      "streaming": "true",
      "historyMaxEntries": "10",
      "allowDnsReverseLookup": "true",
      "maxObjects": "0",
      "debug": "false",
      "serializeException": "false",
      "multicastGroup": "239.192.48.84",
      "maxDepth": "15",
      "authMode": "basic",
      "authMatch": "any",
      "canonicalNaming": "true",
      "allowErrorDetails": "true",
      "realm": "jolokia",
      "includeStackTrace": "true",
      "multicastPort": "24884",
      "useRestrictorService": "false",
      "debugMaxEntries": "100"
    },
    "info": {}
  },
  "timestamp": 1711543870,
  "status": 200
}
```

Jolokia 2 disables discovery by default: https://github.com/jolokia/jolokia/blame/v2.0.2/agent/jvm/src/main/resources/default-jolokia-agent.properties#L75
Jolokia 1 has it enabled: https://github.com/jolokia/jolokia/blob/v1.7.2/agent/jvm/src/main/resources/default-jolokia-agent.properties#L75

**Comment by tadayosi** [comment]
@grgrzybek Thanks for the great investigation!  So it appears that this issue is something that should be patched at camel-jbang. The change you made at the jvm agent between 1.7.x -> 2.0.x makes sense, as it should be safer to disable discovery by default. Camel JBang should just explicitly turn it on to keep the consistent behaviour. I think it's ok for Camel JBang as it has been turned on.

Would you send a pull req to camel-jbang?  We should send it to both `main` and `camel-4.4.x`.

**Comment by tadayosi** [comment]
Fixed at Camel JBang: apache/camel#13645
